### PR TITLE
Slide Matching Algorithm Improvements

### PR DESCRIPTION
**Pull request by kwakmeensoo**
# Slide Matching Algorithm Improvements

## 📝 Description

This PR introduces significant improvements to the slide matching algorithm through the addition of a minimum sentence length filter and optimized hyperparameters.

**Problem**: Short sentences (e.g., "Yes", "Okay", "Um") were being matched to slides based on semantic similarity, often causing incorrect slide transitions and disrupting the overall accuracy of the slide-sentence alignment.

**Solution**:
1. Added a `min_sentence_length` parameter that filters out short sentences from similarity-based matching, allowing them to inherit the previous sentence's slide assignment through the DP algorithm
2. Conducted hyperparameter search to optimize matching parameters for better overall performance

This improvement ensures that only substantive sentences drive slide transitions, while short utterances naturally follow the context of preceding sentences.

---

### 1. Minimum Sentence Length Feature
- **[slide_matching_processor.py](backend/inference_models/slide_matching_processor.py)**: Added `min_sentence_length` parameter
  - Counts words in each sentence during DP matching
  - Short sentences (< threshold) receive zero similarity score, relying only on jump penalty
  - Context scores skip updates for short sentences to maintain semantic coherence
- **[lecture_pipeline.py](backend/inference_models/lecture_pipeline.py)**: Propagated `min_sentence_length` parameter through pipeline initialization
- **[main.py](backend/main.py)**: Configured production value `min_sentence_length = 2` based on testing

### 2. Hyperparameter Optimization
Updated production parameters in [main.py](backend/main.py) based on hyperparameter search results:
- `backward_weight`: 2.0 → 1.85 (reduced penalty for backward jumps)
- `exponential_scale`: 2.79 → 2.785 (fine-tuned confidence scaling)
- `confidence_threshold`: 0.9 → 0.913 (stricter threshold for high-confidence matches)
- `confidence_weight`: 2.13 → 2.18 (increased weight for confident predictions)
- `context_weight`: 0.047 → 0.04 (slightly reduced context influence)

### 3. Technical Implementation Details
- Word counting via `len(sentence.split())` for efficient length checks
- Zero-score assignment for short sentences ensures DP algorithm handles transitions naturally
- Context score EMA updates skip short sentences to preserve semantic context
- Backward compatibility: `min_sentence_length = 0` maintains existing behavior

---

## ✅ Checklist
- [x] Code compiles without errors
- [x] All tests passed locally (tested with `test.py`, `test_korean.py`, `test_male_voice.py`)
- [x] Added/updated tests (N/A - algorithm improvement, existing test files validate)
- [x] Updated relevant documentation (parameters documented in docstrings)
- [x] Followed coding style and conventions

---

### Example Scenario
**Before**: Short sentences like "Um", "Yeah", "Okay" would match to random slides based on spurious semantic similarity.

**After**: Short sentences inherit the slide number from the previous substantive sentence, maintaining context and reducing incorrect transitions.

**Configuration**:
```python
min_sentence_length = 2  # Sentences with < 2 words use zero similarity score
```

---

## 🔗 Related Issues / PRs

This PR builds upon the slide matching feature and addresses accuracy issues observed during testing and production use.

---

## 👥 Reviewers

- @soarhigh03 
- @del2090 

---

### CODE REVIEW

**Review by soarhigh03 — APPROVED** [review]
lgtm

**Review by Soo-03 — APPROVED** [review]
LGTM
